streamingccl: Implement stream_partition builtin.

miretskiy opened the pull request:
Implement `crdb_internal.stream_partition` builtin to emit
partition data during replication streaming.

Fixes #70924

Release Notes: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73685)
<!-- Reviewable:end -->

miretskiy commented:
Only the top commit is about stream_partition builtin.  The rest of the code is being reviewed in https://github.com/cockroachdb/cockroach/pull/71672

Keeping this as a draft PR since I want to expand test coverage for this code; but I think it's a good first start.
Please comment, let me know if you see any glaring issues.

gh-casper reviewed (changes requested):
> Reviewed 9 of 9 files at r1, 17 of 17 files at r2, 1 of 1 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, and @samiskin)

___
*[-- commits, line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-MqsygQ80Mb1XDdGu1pK):*
> ```
> None
> ```
So all crdb_internal builtins don't need release notes?
___
*[pkg/ccl/streamingccl/replication_manager.go, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt1iPY5LCjE3bh1pfC:-Mqt1iPY5LCjE3bh1pfD:b-7ie2uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/replication_manager.go#L67)):*
> ```
> opaqueSpec []byte,
> ```
Just need you to agree on this:

I think the client side can't always treat subscription token as opaque bytes after the planning.

Let's say clients gets a partition spec starting from t1 after it calls replication_stream_spec, it starts consuming the stream with the subscription token until the streams stops at t2 because of a network error. It will restart the stream processor (by calling stream_partition) with a new start_from, t2. Client has to convert the previous subscription token to streampb.StreamPartitionSpec and update the start_from. 

The client shouldn't replan the whole stream because of a transient network error on a partition.
___
*[pkg/ccl/streamingccl/replication_manager.go, line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt5smc8O3lJcXH9EvZ:-Mqt5smc8O3lJcXH9Ev_:b-ctwyxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/replication_manager.go#L74)):*
> ```
> opaqueSpec
> ```
Currently the streampb.StreamPartitionSpec doesn't support backfill checkpoints yet. Or you want to add extend the function signature to include checkpointToken later? It looks like you want to support backfill checkpoints in the next PR? If so, can you add a TODO here?
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 102 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt6qFAAyTkWzqnFY0l:-Mqt6qFAAyTkWzqnFY0m:bcfdw8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L102)):*
> ```
> partitionStreamDecoder
> ```
It looks like the partitionedStreamClient, which I'm working on right now, also needs this to decode row to  streampb.StreamEvent. Can you move the decoder library to streamproducer from streamproducer_test?
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt7y5n09WJr9pSKI2T:-Mqt7y5n09WJr9pSKI2U:bwxb0gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L358)):*
> ```
> StreamPartitionSpec_ExecutionConfig
> ```
I'm wondering if, on the client side, we should get there settings from a stream ingestion sql statement or some cluster settings?


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> Reviewed 1 of 13 files at r5, 3 of 31 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, and @samiskin)

___
*[-- commits, line 55 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-MqsygQ80Mb1XDdGu1pK):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> ```
> None
> ```
So all crdb_internal builtins don't need release notes?
</blockquote></details>

Not really; but in this case this builtin cannot be used as a feature; and it shouldn't really be used directly from SQL shell either.
When we make replication generally available, I suspect we'd include release notes then.
___
*[pkg/ccl/streamingccl/replication_manager.go, line 67 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt1iPY5LCjE3bh1pfC:-MqtNkCBDqRNMyEd2O3m:b-ie8oy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/replication_manager.go#L67)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> ```
> opaqueSpec []byte,
> ```
Just need you to agree on this:

I think the client side can't always treat subscription token as opaque bytes after the planning.

Let's say clients gets a partition spec starting from t1 after it calls replication_stream_spec, it starts consuming the stream with the subscription token until the streams stops at t2 because of a network error. It will restart the stream processor (by calling stream_partition) with a new start_from, t2. Client has to convert the previous subscription token to streampb.StreamPartitionSpec and update the start_from. 

The client shouldn't replan the whole stream because of a transient network error on a partition.

</blockquote></details>

Well, you're right that this spec cannot be treated as opaque.  The client library definitely needs
to be aware of times.  Having said that, this method is invoked via crdb_internal.partition stream and that function gets an argument "[]byte" which it then deserializes (so, it stops being opaque)... I could rename this argument to something generic, like `data` if you prefer.

Regarding the client replanning the whole flow or not -- this is purely an optimization, and does not effect correctness.
___
*[pkg/ccl/streamingccl/replication_manager.go, line 74 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt5smc8O3lJcXH9EvZ:-MqtO9mY2AFndrj1wewv:bhp0yps) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/replication_manager.go#L74)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> ```
> opaqueSpec
> ```
Currently the streampb.StreamPartitionSpec doesn't support backfill checkpoints yet. Or you want to add extend the function signature to include checkpointToken later? It looks like you want to support backfill checkpoints in the next PR? If so, can you add a TODO here?

</blockquote></details>

No, I don't want to extend it.  The checkpoint token will come in in the "opaqueSpec" which we deserialize... It can contain checkpoint token.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 102 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt6qFAAyTkWzqnFY0l:-MqtOMPC0gZI8by9pDbH:b-ywhfjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L102)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> ```
> partitionStreamDecoder
> ```
It looks like the partitionedStreamClient, which I'm working on right now, also needs this to decode row to  streampb.StreamEvent. Can you move the decoder library to streamproducer from streamproducer_test?
</blockquote></details>

I think this is a test only implementation.  I'm depending on testing.T.  This code should not be used in the
not test code.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 358 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mqt7y5n09WJr9pSKI2T:-MqtOarX13DHwXYK6aG2:b7wzaad) ([raw file](https://github.com/cockroachdb/cockroach/blob/149dcfaa8e343cd4562e491c8330b74e14163e8d/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L358)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

> ```
> StreamPartitionSpec_ExecutionConfig
> ```
I'm wondering if, on the client side, we should get there settings from a stream ingestion sql statement or some cluster settings?
</blockquote></details>

Yes, somehow the config should be populated with reasonable values -- either from the sql statement or cluster setting as you say... Alternatively, it's perfectly fine not to do anything right now -- I think I picked somewhat reasonable defaults.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@gh-casper @dt would appreciate some feedback.

dt commented on pkg/ccl/streamingccl/replication_manager.go:
```diff
@@ -61,3 +61,20 @@ func (r replicationStreamManagerImpl) UpdateReplicationStreamProgress(
 }
+
+// StreamPartitionHook hooks a StreamPartition implementation inside streamingccl package.
+var StreamPartitionHook func(
```
I don't really understand why we have all these hooks and mutable function pointers in the ccl code.

miretskiy replied:
yeah, I think we need to clean this up.  There should be just 1 hook to set the manager.  Though, maybe the issue was w/ circular dependencies on jobs package? @gh-casper do you remember?

dt commented on pkg/sql/conn_executor_exec.go:
```diff
@@ -970,7 +970,7 @@ func (ex *connExecutor) dispatchToExecutionEngine(
 	// Certain statements want their results to go to the client
 	// directly. Configure this here.
-	if planner.curPlan.avoidBuffering {
+	if planner.curPlan.avoidBuffering || ex.sessionData().AvoidBuffering {
```
I don't really have much expertise in the sql session / serialize session / exec stuff -- my weak preference would be to split out the avoid buffering bits to their own PR that someone on SQL could stamp too, but it's a weak preference.

miretskiy replied:
Yeah; I spoke w/ Yahor on this.  I just want to have this together w/ the use case for this setting.
@yuzefovich do you mind taking a peak at the AvoidBuffering bit?

dt reviewed (commented):
still working my way through event_stream.go but two quick comments on the rest.

gh-casper reviewed (changes requested):
> Reviewed 9 of 13 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @miretskiy, and @samiskin)

___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 206 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4lKI72QVpiBc2tlXy:-Mr4lKI8-EjbpLRQGXoS:bklmuxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/event_stream.go#L206)):*
> ```Go
> 
> // Close implements tree.ValueGenerator interface.
> func (s *eventStream) Close(ctx context.Context) {
> ```

I'm wondering if this Close will be called when the PG connection with the client times out? Imagine a scenario that the network is slow, client keeps recreating a new `eventStream` on the same node while the old `eventStream` is not GCed, then we will OOM that node.



_Code quote:_
```Go
Close
```
___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 299 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4vqKUCJ6SSWVFUlBS:-Mr4vqKUCJ6SSWVFUlBT:b-xr3pi3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/event_stream.go#L299)):*
> ```Go
> 
> 	// Handle previously skipped updates.
> 	if p.skipped {
> ```

Can you explain why we track `skipped` as code comment?



_Code quote:_
```Go
p.skipped
```
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 117 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4ydC56CHKt0tvBwbL:-Mr4ydC56CHKt0tvBwbM:b-e18b4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L117)):*
> ```Go
> func (d *partitionStreamDecoder) pop() streamingccl.Event {
> 	if d.e.Checkpoint != nil {
> 		// TODO(yevgeniy): Fix checkpoint handling and support backfill checkpoints.
> ```

What needs to be done for checkpoint handling and backfill checkpoints support, in the context of pop?



_Code quote:_
```Go
TODO(yevgeniy): Fix checkpoint handling and support backfill checkpoints.
```
___
*[pkg/kv/kvclient/rangefeed/scanner.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr52yBq6AO3dBukMQrE:-Mr52yBq6AO3dBukMQrF:bwg10jt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/kv/kvclient/rangefeed/scanner.go#L103)):*
> ```Go
> 	}
> 
> 	var fm syncutil.Mutex
> ```

We can move this variable into the functor where it's used.



_Code quote:_
```Go
var fm syncutil.Mutex
```
___
*[pkg/kv/kvclient/rangefeed/scanner.go, line 120 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr53S4Q3K7_fhqOkzXa:-Mr53S4Q3K7_fhqOkzXb:bjisnlz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/kv/kvclient/rangefeed/scanner.go#L120)):*
> ```Go
> 	var retrySpans []roachpb.Span
> 	return func() []roachpb.Span {
> 		if !isRetry {
> ```

Wait. Here `!isRetry` is always true, so we always return all the spans?



_Code quote:_
```Go
if !isRetry {
```


<!-- Sent from Reviewable.io -->

miretskiy commented on pkg/sql/sem/builtins/replication_builtins.go:
```diff
@@ -135,4 +135,34 @@ var replicationBuiltins = map[string]builtinDefinition{
+				[]*types.T{types.Bytes},
+				[]string{"stream_event"},
+			),
+			func(ctx *tree.EvalContext, args tree.Datums) (tree.ValueGenerator, error) {
```
Yeah; looks like we have to do it for all calls here.  I'm going to address this in another PR.

miretskiy reviewed (commented):
Ack.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-:-Mr7Y1TK6z5VXTZaVHfz:bb6sh2i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @yuzefovich)

___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 206 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4lKI72QVpiBc2tlXy:-Mr7_pQTCtliuQsAtuEH:b-vsfd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/event_stream.go#L206)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

I'm wondering if this Close will be called when the PG connection with the client times out? Imagine a scenario that the network is slow, client keeps recreating a new `eventStream` on the same node while the old `eventStream` is not GCed, then we will OOM that node.
</blockquote></details>

I think it's fine.  First, we shouldn't OOM because execution of this code is tied to SQL memory monitor, which we also use here to reserve memory.
Secondly, if we started running this code (i.e. Start() didn't return an error), the ValueGenerator API indicates that Close will be called.
___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 299 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4vqKUCJ6SSWVFUlBS:-Mr7_R1DCm7NQVpjvHUq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/event_stream.go#L299)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Can you explain why we track `skipped` as code comment?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/streamproducer/replication_stream_test.go, line 117 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr4ydC56CHKt0tvBwbL:-Mr7Yfjl9RuYsAUgzZrV:b-im1p7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/streamproducer/replication_stream_test.go#L117)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

What needs to be done for checkpoint handling and backfill checkpoints support, in the context of pop?
</blockquote></details>

streamingccl.Event assumes that checkpoint is a single timestamp.
___
*[pkg/kv/kvclient/rangefeed/scanner.go, line 103 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr52yBq6AO3dBukMQrE:-Mr7Z8eBAwPcspZ-DjCZ:bxlz7dl) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/kv/kvclient/rangefeed/scanner.go#L103)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

We can move this variable into the functor where it's used.
</blockquote></details>

This code is merged already; and now, we can't move it inside the functor because the purpose of this lock
is to protect against concurrent invocations of the function by the parallel scan logic.
___
*[pkg/kv/kvclient/rangefeed/scanner.go, line 120 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr53S4Q3K7_fhqOkzXa:-Mr7ZKUD8w5J0NZ-SelP:bma2jid) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/kv/kvclient/rangefeed/scanner.go#L120)):*
<details><summary><i>Previously, gh-casper wrote…</i></summary><blockquote>

Wait. Here `!isRetry` is always true, so we always return all the spans?
</blockquote></details>

No, that's not correct.  !isRetry is true only the first time this functor executes.  The isRetry is closed over by this functor and I assign this functor to `getSpansToScan := f.getSpansToScan(ctx)` once.  Arguably, this can be made clearer somehow... perhaps by passing in a boolean* to this function.  If you feel strongly -- this can be cleaned up in a separate PR (since this one already merged).


<!-- Sent from Reviewable.io -->

miretskiy commented:
@yuzefovich mind taking a look at the first commit (avoid_buffering session var)?

Commit:
streamingccl: Implement stream_partition builtin.

Implement `crdb_internal.stream_partition` builtin to emit
partition data during replication streaming.

Release Notes: None

yuzefovich reviewed (commented):
One question on the first commit from me.

> Reviewed 1 of 13 files at r8, 3 of 4 files at r10, 4 of 13 files at r11, 11 of 11 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-:-Mr94IWEBEub6Kduer_N:b-e68n98)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 639 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr942LwC_MUeNMRVuYZ:-Mr942LwC_MUeNMRVuY_:b-b6olvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/de72b264dd906510b48d2831f7c1a66da59bf073/pkg/sql/logictest/testdata/logic_test/crdb_internal#L639)):*
> ```
> ----
> start_pretty   end_pretty
> /Table/62/1/1  /Table/62/1/2
> ```

Why this change is in this commit?


<!-- Sent from Reviewable.io -->

dt commented on pkg/ccl/streamingccl/streamproducer/event_stream.go:
```diff
@@ -0,0 +1,438 @@
+	if err != nil {
+		return err
+	}
+
```
I'm trying to figure out if I think we should be boxing these bytes in another proto, to add any sort of versioning or anything, leave room for a compression-identifying header or whatever.

I suppose we could just say that the passed spec would ask for a particular stream version / encoding / compression / etc that it expects to decode and then this producer is just obligated to either produce what is requested or refuse with an error, so then don't need to worry about it identifying what we produced in the produced data?

so just leaving this as a naked marshaled proto would be okay?

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @samiskin, and @yuzefovich)

___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 269 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mq_FhbbQNscopQuPUbp-r14-269:-MrCjpi3AZ3v4MvGAjS-:bpecr99) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0098106d3ce072e469b1470ab35c09f47f4256/pkg/ccl/streamingccl/streamproducer/event_stream.go#L269)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'm trying to figure out if I think we should be boxing these bytes in another proto, to add any sort of versioning or anything, leave room for a compression-identifying header or whatever.

I suppose we could just say that the passed spec would ask for a particular stream version / encoding / compression / etc that it expects to decode and then this producer is just obligated to either produce what is requested or refuse with an error, so then don't need to worry about it identifying what we produced in the produced data?

so just leaving this as a naked marshaled proto would be okay?
</blockquote></details>

We could add more info to StreamEvent itself, can we?
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 639 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr942LwC_MUeNMRVuYZ:-MrCjYlEAa1Q0P2LnzHb:bmv7z7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/de72b264dd906510b48d2831f7c1a66da59bf073/pkg/sql/logictest/testdata/logic_test/crdb_internal#L639)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why this change is in this commit?
</blockquote></details>

I actually don't know... I just ran logic test w/ rewrite -- this was picked up.
I can try backing out this change to see if tests still pass... Thoughts?


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @samiskin, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 639 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr942LwC_MUeNMRVuYZ:-MrCnyQ193nMbmb1vMFJ:bgiobml) ([raw file](https://github.com/cockroachdb/cockroach/blob/de72b264dd906510b48d2831f7c1a66da59bf073/pkg/sql/logictest/testdata/logic_test/crdb_internal#L639)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I actually don't know... I just ran logic test w/ rewrite -- this was picked up.
I can try backing out this change to see if tests still pass... Thoughts?
</blockquote></details>

I reran on master -- got same rewrite data.  


<!-- Sent from Reviewable.io -->

Commit:
sql: Add avoid_buffering session variable.

Add `avoid_buffering` session variable which turns off buffering
at the connection executor level.  This sesson variable is needed
so that streaming replication can manage its own flushing behavior.

Release Notes: None

miretskiy reviewed (commented):
@yuzefovich I repushed changes with seemingly unrelated logic test rewrites removed; though I'm not sure why those also get rewritten on the master.  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-:-MrCt2CbAD5LrW_P0_0I:btygj5y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @samiskin, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 639 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr942LwC_MUeNMRVuYZ:-MrCsxk82kuZux8yNBa-:bqykpf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/de72b264dd906510b48d2831f7c1a66da59bf073/pkg/sql/logictest/testdata/logic_test/crdb_internal#L639)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I reran on master -- got same rewrite data.  
</blockquote></details>

I re-pushed changes with seeming


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @yuzefovich)

___
*[pkg/ccl/streamingccl/streamproducer/event_stream.go, line 269 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mq_FhbbQNscopQuPUbp-r14-269:-MrD1Q1U0crZKvtuINAL:b-5kh6t8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a0098106d3ce072e469b1470ab35c09f47f4256/pkg/ccl/streamingccl/streamproducer/event_stream.go#L269)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

We could add more info to StreamEvent itself, can we?
</blockquote></details>

Not if we want to compress the marshaled streamevent itself right? but we could make a new field of streamevent that is just bytes that hold a compressed, marshaled inner streamevent or something I guess, so between that or just saying the client gets to specify in the spec what is in these bytes (compressed or not), I'm convinced we have options so don't need to do anything right now.


<!-- Sent from Reviewable.io -->

miretskiy commented:
Ack; good point re compression - -I forget we're not running under grpc here... But, as you say, we have options, so going to keep it simple for now.

miretskiy commented:
@yuzefovich Looks like backing out those few changes still has CI clean.

yuzefovich reviewed (approved):
First commit <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a nit.

> Reviewed 11 of 11 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-:-MrNtGMl8J5TN-1ikJ29:b-hzu902)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @yuzefovich)

___
*[pkg/sql/sessiondatapb/local_only_session_data.proto, line 218 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-MrNsx6e4Bs_qDLJay6C:-MrNsx6e4Bs_qDLJay6D:b-dglire) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c28dad68f1f961a52ee17516898d3f065862341/pkg/sql/sessiondatapb/local_only_session_data.proto#L218)):*
> ```protobuf
>   // recommendations below the plan in EXPLAIN.
>   bool index_recommendations_enabled = 58;
> ```

nit: looks like we're not adding new lines here before each new field.
___
*[pkg/sql/logictest/testdata/logic_test/crdb_internal, line 639 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mr942LwC_MUeNMRVuYZ:-MrNtBXr2chUia9ZK8o9:b9l4m9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/de72b264dd906510b48d2831f7c1a66da59bf073/pkg/sql/logictest/testdata/logic_test/crdb_internal#L639)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I re-pushed changes with seeming
</blockquote></details>

Hm, I'm guessing it's because of `retry` option.


<!-- Sent from Reviewable.io -->

gh-casper reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73685)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dt, @gh-casper, @miretskiy, @samiskin, and @yuzefovich)

___
*[pkg/ccl/streamingccl/replication_manager.go, line 66 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/73685#-Mq_FhbbQNscopQuPUbf-r9-66:-MrO-jnd4VW0sGdVpxPU:b-e1bkyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa01f8dcda3c890b335efba50ecbb08258e441dd/pkg/ccl/streamingccl/replication_manager.go#L66)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

yeah, I think we need to clean this up.  There should be just 1 hook to set the manager.  Though, maybe the issue was w/ circular dependencies on jobs package? @gh-casper do you remember?
</blockquote></details>

Yes, the issue was circular dependencies on some packages inside streamccl such as streamccl/streamproducer. But it's good to see you solve this in the other PR.


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+
